ecs-cli/modules/cli/local/network/setup.go

```diff
@@ -122,6 +132,53 @@ func createLocalNetwork(dockerClient networkCreator) {
 	logrus.Infof("Created network %s with ID %s", EcsLocalNetworkName, resp.ID)
 }
 
+func setupLocalEndpointsImage(dockerClient imagePuller) {
+	if localEndpointsImageExists(dockerClient) {
+		return
+	}
+	pullLocalEndpointsImage(dockerClient)
+}
+
+func localEndpointsImageExists(dockerClient imagePuller) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), docker.TimeoutInS)
+	defer cancel()
+
+	args := filters.NewArgs(filters.Arg("reference", localEndpointsImageName))
+	imgs, err := dockerClient.ImageList(ctx, types.ImageListOptions{
+		Filters: args,
+	})
+	if err != nil {
+		logrus.Fatalf("Failed to list images with filters %v due to %v", args, err)
+	}
+
+	for _, img := range imgs {
+		for _, repotag := range img.RepoTags {
+			if strings.HasPrefix(repotag, localEndpointsImageName) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func pullLocalEndpointsImage(dockerClient imagePuller) {
+	ctx, cancel := context.WithTimeout(context.Background(), docker.LongTimeoutInS)
+	defer cancel()
+
+	logrus.Infof("Pulling image %s", localEndpointsImageName)
+	rc, err := dockerClient.ImagePull(ctx, localEndpointsImageName, types.ImagePullOptions{})
```

Comment:
Does ImagePull download the image already? If that's the case, maybe we don't need to load the entire image from persistence storage into memory again on ln 175.

Reply:
Good question! Unfortunately nope :/ ImagePull returns immediately the moment that the download starts. We need to read the full stream to make sure that the download completes without an error.  

Instead of `ReadAll` I could read into a fixed size byte array and keep replacing its content until `io.EOF` but I don't think it's worth it :P

Reply:
Another newbie question, seems like even though we `ReadAll` on `rc`, we never persist the bytes read, so they  must not be the image we're pulling, correct?

Reply:
The stream contains only status updates of the image we're pulling. Here is the output of `io.Copy(os.Stdout, rc)`:
```shell
{"status":"Pulling from amazon/amazon-ecs-local-container-endpoints","id":"latest"}
{"status":"Pulling fs layer","progressDetail":{},"id":"eb2aeee9dc99"}
{"status":"Pulling fs layer","progressDetail":{},"id":"e8bad1a936f2"}
{"status":"Pulling fs layer","progressDetail":{},"id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":16360,"total":120766},"progress":"[======\u003e                                            ]  16.36kB/120.8kB","id":"eb2aeee9dc99"}
{"status":"Downloading","progressDetail":{"current":8239,"total":8239},"progress":"[==================================================\u003e]  8.239kB/8.239kB","id":"e8bad1a936f2"}
{"status":"Verifying Checksum","progressDetail":{},"id":"e8bad1a936f2"}
{"status":"Download complete","progressDetail":{},"id":"e8bad1a936f2"}
{"status":"Downloading","progressDetail":{"current":68139,"total":120766},"progress":"[============================\u003e                      ]  68.14kB/120.8kB","id":"eb2aeee9dc99"}
{"status":"Verifying Checksum","progressDetail":{},"id":"eb2aeee9dc99"}
{"status":"Download complete","progressDetail":{},"id":"eb2aeee9dc99"}
{"status":"Extracting","progressDetail":{"current":32768,"total":120766},"progress":"[=============\u003e                                     ]  32.77kB/120.8kB","id":"eb2aeee9dc99"}
{"status":"Extracting","progressDetail":{"current":120766,"total":120766},"progress":"[==================================================\u003e]  120.8kB/120.8kB","id":"eb2aeee9dc99"}
{"status":"Downloading","progressDetail":{"current":68096,"total":5653304},"progress":"[\u003e                                                  ]   68.1kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Pull complete","progressDetail":{},"id":"eb2aeee9dc99"}
{"status":"Downloading","progressDetail":{"current":137728,"total":5653304},"progress":"[=\u003e                                                 ]  137.7kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Extracting","progressDetail":{"current":8239,"total":8239},"progress":"[==================================================\u003e]  8.239kB/8.239kB","id":"e8bad1a936f2"}
{"status":"Extracting","progressDetail":{"current":8239,"total":8239},"progress":"[==================================================\u003e]  8.239kB/8.239kB","id":"e8bad1a936f2"}
{"status":"Downloading","progressDetail":{"current":207360,"total":5653304},"progress":"[=\u003e                                                 ]  207.4kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Pull complete","progressDetail":{},"id":"e8bad1a936f2"}
{"status":"Downloading","progressDetail":{"current":276992,"total":5653304},"progress":"[==\u003e                                                ]    277kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":346624,"total":5653304},"progress":"[===\u003e                                               ]  346.6kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":485888,"total":5653304},"progress":"[====\u003e                                              ]  485.9kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":625152,"total":5653304},"progress":"[=====\u003e                                             ]  625.2kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":764416,"total":5653304},"progress":"[======\u003e                                            ]  764.4kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":903680,"total":5653304},"progress":"[=======\u003e                                           ]  903.7kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":1042944,"total":5653304},"progress":"[=========\u003e                                         ]  1.043MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":1182208,"total":5653304},"progress":"[==========\u003e                                        ]  1.182MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":1321472,"total":5653304},"progress":"[===========\u003e                                       ]  1.321MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":1460736,"total":5653304},"progress":"[============\u003e                                      ]  1.461MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":1669632,"total":5653304},"progress":"[==============\u003e                                    ]   1.67MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":1878528,"total":5653304},"progress":"[================\u003e                                  ]  1.879MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":2017792,"total":5653304},"progress":"[=================\u003e                                 ]  2.018MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":2226688,"total":5653304},"progress":"[===================\u003e                               ]  2.227MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":2435584,"total":5653304},"progress":"[=====================\u003e                             ]  2.436MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":2644480,"total":5653304},"progress":"[=======================\u003e                           ]  2.644MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":2923008,"total":5653304},"progress":"[=========================\u003e                         ]  2.923MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":3201536,"total":5653304},"progress":"[============================\u003e                      ]  3.202MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":3549696,"total":5653304},"progress":"[===============================\u003e                   ]   3.55MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":3897856,"total":5653304},"progress":"[==================================\u003e                ]  3.898MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":4176384,"total":5653304},"progress":"[====================================\u003e              ]  4.176MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":4594176,"total":5653304},"progress":"[========================================\u003e          ]  4.594MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":4942336,"total":5653304},"progress":"[===========================================\u003e       ]  4.942MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Downloading","progressDetail":{"current":5429760,"total":5653304},"progress":"[================================================\u003e  ]   5.43MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Verifying Checksum","progressDetail":{},"id":"b334ff5ac2f9"}
{"status":"Download complete","progressDetail":{},"id":"b334ff5ac2f9"}
{"status":"Extracting","progressDetail":{"current":65536,"total":5653304},"progress":"[\u003e                                                  ]  65.54kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Extracting","progressDetail":{"current":327680,"total":5653304},"progress":"[==\u003e                                                ]  327.7kB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Extracting","progressDetail":{"current":2621440,"total":5653304},"progress":"[=======================\u003e                           ]  2.621MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Extracting","progressDetail":{"current":3866624,"total":5653304},"progress":"[==================================\u003e                ]  3.867MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Extracting","progressDetail":{"current":5653304,"total":5653304},"progress":"[==================================================\u003e]  5.653MB/5.653MB","id":"b334ff5ac2f9"}
{"status":"Pull complete","progressDetail":{},"id":"b334ff5ac2f9"}
{"status":"Digest: sha256:57b8299abb902c31171450ec013855f6d15116c00ca2108c071be0d8993daf5e"}
{"status":"Status: Downloaded newer image for amazon/amazon-ecs-local-container-endpoints:latest"}
```